Source/WebCore/css/parser/CSSPropertyParser.cpp

```diff
@@ -5745,11 +5753,24 @@ bool CSSPropertyParser::consumeBorderImage(CSSPropertyID property, bool importan
             return true;
         case CSSPropertyBorderImage:
         case CSSPropertyWebkitBorderImage:
-            addPropertyWithImplicitDefault(CSSPropertyBorderImageSource, property, WTFMove(source), valuePool.createImplicitInitialValue(), important);
-            addPropertyWithImplicitDefault(CSSPropertyBorderImageSlice, property, WTFMove(slice), valuePool.createImplicitInitialValue(), important);
-            addPropertyWithImplicitDefault(CSSPropertyBorderImageWidth, property, WTFMove(width), valuePool.createImplicitInitialValue(), important);
-            addPropertyWithImplicitDefault(CSSPropertyBorderImageOutset, property, WTFMove(outset), valuePool.createImplicitInitialValue(), important);
-            addPropertyWithImplicitDefault(CSSPropertyBorderImageRepeat, property, WTFMove(repeat), valuePool.createImplicitInitialValue(), important);
+            addPropertyWithImplicitDefault(CSSPropertyBorderImageSource, property, WTFMove(source), valuePool.createIdentifierValue(CSSValueNone), important);
+            if (slice)
+                addProperty(CSSPropertyBorderImageSlice, property, slice.releaseNonNull(), important);
+            else {
+                auto value = CSSBorderImageSliceValue::create(createQuad(100, CSSUnitType::CSS_PERCENTAGE), property == CSSPropertyWebkitBorderImage);
+                addProperty(CSSPropertyBorderImageSlice, property, WTFMove(value), important, true);
+            }
+            if (width)
+                addProperty(CSSPropertyBorderImageWidth, property, width.releaseNonNull(), important);
+            else {
+                auto value = CSSBorderImageWidthValue::create(createQuad(1, CSSUnitType::CSS_NUMBER), false);
+                addProperty(CSSPropertyBorderImageWidth, property, WTFMove(value), important, true);
+            }
+            if (outset)
+                addProperty(CSSPropertyBorderImageOutset, property, outset.releaseNonNull(), important);
+            else
+                addProperty(CSSPropertyBorderImageOutset, property, createQuad(0, CSSUnitType::CSS_NUMBER), important, true);
```

Comment:
Why can't we keep using `addPropertyWithImplicitDefault()` here? It handles this pattern of `if (value) addExplicit() else addImplicit()`

Reply:
We can, but we end up doing a lot of wasted work to create the implicit value that is thrown away whenever the explicit value is non-null. Those quads are complex objects. I don’t have a strong feeling either way.

Reply:
I did it. Code looks much shorter, is less efficient.

Reply:
Maybe `addPropertyWithImplicitDefault` should take a lambda that creates the value if needed instead of the value itself?